openssl-dynamic/src/main/c/ssl.c

```diff
@@ -221,6 +234,243 @@ static const jint supported_ssl_opts = 0
+    return (int) tcn_write_to_bytebuffer(bio, in, inl);
+}
+
+static int bio_java_bytebuffer_read(BIO* bio, char* out, int outl) {
+    BIO_clear_retry_flags(bio);
+    return (int) tcn_read_from_bytebuffer(bio, out, outl);
+}
+
+static int bio_java_bytebuffer_puts(BIO* bio, const char *in) {
+    BIO_clear_retry_flags(bio);
+    return (int) tcn_write_to_bytebuffer(bio, in, strlen(in));
+}
+
+static int bio_java_bytebuffer_gets(BIO* b, char* out, int outl) {
+    // Not supported https://www.openssl.org/docs/man1.0.2/crypto/BIO_write.html
+    return -2;
+}
+
+static long bio_java_bytebuffer_ctrl(BIO* bio, int cmd, long num, void* ptr) {
+    // see https://www.openssl.org/docs/man1.0.1/crypto/BIO_ctrl.html
+    switch (cmd) {
+        case BIO_CTRL_GET_CLOSE:
+            return (long) BIO_get_shutdown(bio);
+        case BIO_CTRL_SET_CLOSE:
+            BIO_set_shutdown(bio, (int) num);
+            return 1;
+        case BIO_CTRL_FLUSH:
+            return 1;
+        case BIO_CTRL_PENDING: {
+            struct TCN_bio_bytebuffer* bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio);
+            return bioUserData == NULL ? 0 : bioUserData->bufferLength;
+        }
+        case BIO_CTRL_WPENDING: {
+            struct TCN_bio_bytebuffer* bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio);
+            return bioUserData == NULL ? 0 : bioUserData->sslWriteBufferLength;
+        }
+        default:
+            return 0;
+    }
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
+static BIO_METHOD bio_java_bytebuffer_methods = {
+    BIO_TYPE_MEM,
+    "Java ByteBuffer",
+    bio_java_bytebuffer_write,
+    bio_java_bytebuffer_read,
+    bio_java_bytebuffer_puts,
+    bio_java_bytebuffer_gets,
+    bio_java_bytebuffer_ctrl,
+    bio_java_bytebuffer_create,
+    bio_java_bytebuffer_destroy,
+    NULL
+};
+#else
+static BIO_METHOD* bio_java_bytebuffer_methods = NULL;
+
+static void init_bio_methods(void) {
+    bio_java_bytebuffer_methods = BIO_meth_new(BIO_TYPE_FILE, "Java ByteBuffer");
+    BIO_meth_set_write(bio_java_bytebuffer_methods, &bio_java_bytebuffer_write);
+    BIO_meth_set_read(bio_java_bytebuffer_methods, &bio_java_bytebuffer_read);
+    BIO_meth_set_puts(bio_java_bytebuffer_methods, &bio_java_bytebuffer_puts);
+    BIO_meth_set_gets(bio_java_bytebuffer_methods, &bio_java_bytebuffer_gets);
+    BIO_meth_set_ctrl(bio_java_bytebuffer_methods, &bio_java_bytebuffer_ctrl);
+    BIO_meth_set_create(bio_java_bytebuffer_methods, &bio_java_bytebuffer_create);
+    BIO_meth_set_destroy(bio_java_bytebuffer_methods, &bio_java_bytebuffer_destroy);
+}
+
+static void free_bio_methods(void) {
+    BIO_meth_free(bio_java_bytebuffer_methods);
+}
+#endif
+
+static BIO_METHOD* BIO_java_bytebuffer() {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
+    return &bio_java_bytebuffer_methods;
+#else
+    return bio_java_bytebuffer_methods;
+#endif
```

Comment:
The APIs changed between SSL versions. `BIO_meth_new` and `BIO_meth_free` are newer.